build-system/babel-plugins/babel-plugin-transform-function-declarations/index.js

```diff
@@ -14,145 +14,134 @@
-  isNewedInProgramScope: (t, path) => {
-    const {name} = path.get('id').node;
-    let isNewed = false;
-    path
-      .findParent(path => path.isProgram())
-      .traverse({
-        NewExpression(path) {
-          if (t.isIdentifier(path.node.callee, {name})) {
-            isNewed = true;
+    // If the function contains usage of `arguments`, bail out on modification.
+    containsArgumentsUsage(path) {
+      let containsArgumentsUsage = false;
+      path.traverse({
+        Identifier(path) {
+          if (t.isIdentifier(path.node, {name: 'arguments'})) {
+            containsArgumentsUsage = true;
             path.stop();
           }
         },
       });
-    return isNewed;
-  },
-};
+      return containsArgumentsUsage;
+    },
+
+    // If the function contains a ThisExpression, converting from a FunctionDeclaration to a
+    // VariableDeclaration => VariableDeclarator => ArrowFunctionExpression isn't necessarily valid.
+    // Bail out on modification.
+    containsThisExpression(path) {
+      let containsThis = false;
+      path.traverse({
+        ThisExpression() {
+          containsThis = true;
+          path.stop();
+        },
+      });
+      return containsThis;
+    },
 
-function createVariableDeclaration(t, path) {
-  const {params, body, async, id} = t.cloneNode(path.node);
-  const arrowFunction = t.arrowFunctionExpression(
-    params,
-    body.body[0].argument,
-    async
-  );
-  const declarations = t.variableDeclarator(
-    t.identifier(id.name),
-    arrowFunction
-  );
+    // If the FunctionDeclaration identifier is newed in the scope of this program, bail out on modification.
+    isNewedInProgramScope(path) {
+      const {name} = path.get('id').node;
+      const binding = path.scope.getBinding(name);
+      return binding.referencePaths.some(p => p.parentPath.isNewExpression());
+    },
+  };
 
-  // Making an existing FunctionDeclaration a let VariableDeclaration is safe.
-  // The name was already reserved for the FunctionDeclaration.
-  return t.variableDeclaration('let', [declarations]);
-}
+  function createVariableDeclaration(path) {
+    const {params, body, async, id} = t.cloneNode(path.node);
+    const arrowFunction = t.arrowFunctionExpression(
+      params,
+      body.body[0].argument,
+      async
+    );
+    const declarations = t.variableDeclarator(
+      t.identifier(id.name),
+      arrowFunction
+    );
 
-// Attempt to convert simple single ReturnStatement FunctionDeclarations to ArrowFunctionExpressions.
-// See BAIL_OUT_CONDITIONS for reasons why FunctionDeclarations would not be modified.
-module.exports = function({types: t}) {
-  const DEBUG = false;
-  const REPORT = false;
+    // Making an existing FunctionDeclaration a let VariableDeclaration is safe.
+    // The name was already reserved for the FunctionDeclaration.
+    return t.variableDeclaration('let', [declarations]);
```

Comment:
just for my own knowledge, we just always insert it at the top of the body correct?

Reply:
Top of the nearest body.

https://github.com/ampproject/amphtml/pull/27438/files#diff-015cde67b5314efc655dc94576b26ce6R134